Add Cockpit role with default variables and tasks

RaneyDazed opened the pull request:
This pull request adds the Cockpit role to the sandbox.yml file and includes the necessary default variables and tasks for the role.

RaneyDazed commented:
Ok. So its not quite ready yet, I was hoping you could take a look and give me an idea of what the hell I did that wasn't up to the salty standard. I left some comments on purpose. Just be nice 😅

RaneyDazed commented:
fwiw, it works fine. I just know that its not perfect. :p

saltydk commented on roles/cockpit/tasks/main.yml:
```diff
@@ -0,0 +1,99 @@
+    name: cockpit.service
+  when: cockpit_service.stat.exists
+
+# Also need to fix the traefik file to include the correct settings
```
Remove the comment.

saltydk commented on roles/cockpit/tasks/main.yml:
```diff
@@ -0,0 +1,99 @@
+- name: Create directories
+  ansible.builtin.include_tasks: "{{ resources_tasks_path }}/directories/create_directories.yml"
+
+- name: Import 'cockpit_socket'
```
Use owner and group fields for permissions.

RaneyDazed replied:
Should they all be owned by user:group? I thought that those would be fine as is with just `mode: 0664`. I figured the only one we'd need to worry about perms on would be the traefik `.yml` file

saltydk commented on roles/cockpit/tasks/main.yml:
```diff
@@ -0,0 +1,99 @@
+    mode: "0664"
+    force: true
+
+- name: Import 'cockpit_traefik'
```
Use owner and group fields for permissions.

saltydk commented on roles/cockpit/tasks/main.yml:
```diff
@@ -0,0 +1,99 @@
+    mode: "0775"
+    force: true
+
+- name: Change traefik file ownership
```
Remove redundant task.

saltydk commented on roles/cockpit/tasks/main.yml:
```diff
@@ -0,0 +1,99 @@
+    group: "{{ gid }}"
+  when:
+    - not continuous_integration
+    # - not cockpit_socket.stat.exists
```
Why is this here?

saltydk commented on roles/cockpit/tasks/main.yml:
```diff
@@ -0,0 +1,99 @@
+    - not continuous_integration
+    # - not cockpit_socket.stat.exists
+
+- name: Install Cockpit from backports
```
Missing state, which should be latest if you ever want this to update and it should also freeze the apt package if you only want the role to manage updates.

saltydk commented on roles/cockpit/tasks/main.yml:
```diff
@@ -0,0 +1,99 @@
+    default_release: "{{ ansible_distribution_release }}-backports"
+    update_cache: yes
+  when:
+    - not cockpit_socket.stat.exists
```
Don't understand why this is here.

saltydk commented on roles/cockpit/templates/cockpit.conf.j2:
```diff
@@ -0,0 +1,5 @@
+[WebService]
+Origins = https://{{ cockpit_web_url }}
```
If you hadn't removed https from the variable it could have been plugged directly in.

saltydk commented on roles/cockpit/templates/cockpit.yml.j2:
```diff
@@ -0,0 +1,45 @@
+        options: "securetls@file"
+      service: "{{ cockpit_name }}"
+
+# actual jinja template for traefik, attempt num 1
```
Not sure why you omitted the http router.

saltydk commented on roles/cockpit/templates/listen.conf.j2:
```diff
@@ -0,0 +1,4 @@
+[Socket]
+ListenStream=
+ListenStream=172.19.0.1:{{ cockpit_web_port }}
```
You will need to make cockpit start after docker, in their systemd services, to listen on this.

saltydk commented:
Probably also worth mentioning that this role does not install the VM bits, so either make that the default or add a toggle to install it along with any other system dependencies.

RaneyDazed commented on roles/cockpit/defaults/main.yml:
```diff
@@ -0,0 +1,62 @@
+cockpit_dns_record: "{{ cockpit_web_subdomain }}"
+cockpit_dns_zone: "{{ cockpit_web_domain }}"
+
+# I removed 'https://' + from the beginning of line #46.
```
Alright, so doesn't the traefik `.yml` file take `sub.domain.tld`? without HTTP or HTTPS? I just threw that together to fit it in there. I wasn't sure if I should do that, or use the combined variables to create it. I'll correct it and use 
```yaml
{{ cockpit_web_subdomain }} '+'. {{ cockpit_web_domain }}
``` 
instead. Or however its formatted.

RaneyDazed replied:
Fixed that.

RaneyDazed commented:
> Probably also worth mentioning that this role does not install the VM bits, so either make that the default or add a toggle to install it along with any other system dependencies.

I'll be working on that after I fix everything else. Going to get plain jain cockpit going, make sure there's no issues from your POV, then I'll mess w the VM portion.

RaneyDazed reviewed (commented):
(Minus the anything VM related) it should be good to go. Overrides in place to not start until docker has started. package is managed by the role. controlled by a setting in defaults.

RaneyDazed commented:
Ok, work in progress VM stuff added. I think I need someone else to test it though. I don't use any VMs right now, and idk what I would do to test it.

saltydk commented on roles/cockpit/tasks/main.yml:
```diff
@@ -0,0 +1,75 @@
+  ansible.builtin.template:
+    src: cockpit.yml.j2
+    dest: "{{ cockpit_paths_traefik_location }}"
+    mode: "0775"
```
Why is a yml file set to executable?

saltydk replied:
Guess it is a copy of what we do, which is also incorrect.

saltydk commented on roles/cockpit/tasks/subtasks/vm.yml:
```diff
@@ -0,0 +1,31 @@
+- name: VM | Install necessary packages for Cockpit Machines
+  ansible.builtin.package:
+    name:
+      - cockpit-machines
```
shouldn't this be from the same backport repo as cockpit?

RaneyDazed replied:
Yes, you're right. I overlooked that. My bad

RaneyDazed commented:
Better?

saltydk commented on roles/cockpit/defaults/main.yml:
```diff
@@ -0,0 +1,68 @@
+cockpit_traefik_enabled: true
+cockpit_traefik_api_enabled: false
+
+################################
+# Settings
+################################
```
Settings block should be after the basics.

saltydk commented on roles/cockpit/tasks/main.yml:
```diff
@@ -0,0 +1,75 @@
+    src: cockpit.yml.j2
+    dest: "{{ cockpit_paths_traefik_location }}"
+    mode: "664"
+    owner: "{{ uid }}"
+    group: "{{ gid }}"
```
Use user.name although it is technically the same it is what the other roles use.

RaneyDazed replied:
You want user.name only, or user.name and user.group or something?

saltydk commented on roles/cockpit/tasks/subtasks/vm.yml:
```diff
@@ -0,0 +1,50 @@
+    default_release: "{{ ansible_distribution_release }}-backports"
+    state: latest
+    update_cache: true
+    allow_downgrade: true
```
Why is this enabled?

saltydk commented on roles/cockpit/tasks/subtasks/package.yml:
```diff
@@ -0,0 +1,42 @@
+    default_release: "{{ ansible_distribution_release }}-backports"
+    state: latest
+    update_cache: true
+    allow_downgrade: true
```
Why is this enabled?

saltydk commented on roles/cockpit/templates/cockpit.yml.j2:
```diff
@@ -0,0 +1,25 @@
+      service: "{{ cockpit_name }}"
+      tls:
+        options: securetls@file
+        certResolver: {{ cockpit_traefik_certresolver}}
```
Fix the formatting.

saltydk commented on roles/cockpit/templates/cockpit.yml.j2:
```diff
@@ -0,0 +1,25 @@
+    {{ cockpit_name }}-http:
+      entryPoints:
+        - "web"
+      rule: "Host(`{{ cockpit_web_subdomain + '.' + cockpit_web_domain }}`)"
```
Maybe try using traefik_host here as it should do the empty subdomain check for you.

saltydk commented on roles/cockpit/templates/cockpit.yml.j2:
```diff
@@ -0,0 +1,25 @@
+    {{ cockpit_name }}:
+      entryPoints:
+        - "websecure"
+      rule: "Host(`{{ cockpit_web_subdomain + '.' + cockpit_web_domain }}`)"
```
Maybe try using traefik_host here as it should do the empty subdomain check for you.

RaneyDazed replied:
Is that defined somewhere? if subdomain isn't defined it'll start with `.` which ain't no good. but I'm not sure, should I be defining traefik_host, or its done by the role somehow I'm unaware of?

saltydk replied:
Yes, it is a global variable.

RaneyDazed commented:
> Maybe try using traefik_host here as it should do the empty subdomain check for you.

It works fine with `traefik_host`.

saltydk commented:
3AM here so I'm gonna pass on reviewing more until tomorrow.

RaneyDazed commented:
> 3AM here so I'm gonna pass on reviewing more until tomorrow.

all good salty! Thanks!

RaneyDazed commented:
need to update it for the network start. everything is great except for that VM task.

saltydk commented on roles/cockpit/tasks/subtasks/vm.yml:
```diff
@@ -0,0 +1,49 @@
+    state: latest
+    update_cache: true
+
+- name: VM | Start default network in libvirt
```
This is why we test stuff as we go instead of relying on ChatGPT for everything.

saltydk commented:
I'm not really going to go over everything as then I'd prefer making the role myself anyway when I get around to doing it. Get someone to test it and if it works well enough for @owine's requirements then merge it.

RaneyDazed commented on roles/cockpit/defaults/main.yml:
```diff
@@ -0,0 +1,68 @@
+cockpit_dns_zone: "{{ cockpit_web_domain }}"
+cockpit_dns_proxy: "{{ dns.proxied }}"
+
+################################
+# Traefik
+################################
+
+cockpit_traefik_sso_middleware: "{{ traefik_default_sso_middleware }}"
+cockpit_traefik_middleware_default: "{{ traefik_default_middleware }}"
+cockpit_traefik_middleware_custom: ""
+cockpit_traefik_certresolver: "{{ traefik_default_certresolver }}"
+cockpit_traefik_enabled: true
```
> Are these doing anything since the traffic handling is done via the file instead of labels? I don't see the middleware vars referenced in the traffic yml template. Should the `cockpit_traefik_enabled` toggle being set to `false` trigger removal of the cockpit.yml from the traefik folder?

hmm. I think so? I'll have to go over the role again to be sure, I used some vars in the yml template but you're prob right. I just stuck them in there without thinking about it. 

oo good. I like that.

but yeah I think that setting it to false should remove the yml template.

RaneyDazed replied:
@owine should be done here! hopefully :p

maximuskowalski commented:
@RaneyDazed

RaneyDazed commented:
> @RaneyDazed 

Sorry max, I started a new job that begins at 6am and I've just been busy. I'll get to it asap I promise.

RaneyDazed commented:
@maximuskowalski when you have a moment, could you take a look? thanks!

owine commented on roles/cockpit/tasks/main.yml:
```diff
@@ -0,0 +1,88 @@
+    dns_record: "{{ lookup('vars', role_name + '_dns_record') }}"
+    dns_zone: "{{ lookup('vars', role_name + '_dns_zone') }}"
+    dns_proxy: "{{ lookup('vars', role_name + '_dns_proxy') }}"
+  when: not cockpit_traefik_enabled
```
Shouldn't this be reversed? We want to set the DNS records when Traefik will be enabled

RaneyDazed replied:
Shit you're right. I had my head wrapped around this the wrong way. 😂

owine commented on roles/cockpit/templates/cockpit.yml.j2:
```diff
@@ -0,0 +1,25 @@
+        - "web"
+      rule: "Host(`{{ traefik_host }}`)"
+      middlewares:
+        {{ traefik_default_middleware_http.split(',') | to_nice_yaml | trim | indent(8) }}
```
No SSO or custom middleware. Why not keep the variable defaults as they were previously, but just reference them here when building the template?

RaneyDazed replied:
I'll have to mull it over a bit. I'm much less familiar with the dynamic traefik files than I should be.

saltydk replied:
There is no need to use that file at all.

owine commented on roles/cockpit/templates/cockpit.yml.j2:
```diff
@@ -0,0 +1,25 @@
+        - "websecure"
+      rule: "Host(`{{ traefik_host }}`)"
+      middlewares:
+        {{ traefik_middleware.split(',') | to_nice_yaml | trim | indent(8) }}
```
Same as above

RaneyDazed commented:
@owine I think we're done with this. I copied the qbittorrent stuff, and that should be sufficient according to salty. I tested with 2 diff servers, removed and added sso. Didn't try to add any custom middleware though.
